src/miral/fd_manager.cpp

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright © 2022 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "fd_manager.h"
+#include "mir/main_loop.h"
+
+#include <boost/throw_exception.hpp>
+
+using namespace miral;
+
+FdManager::FdManager()
+{
+}
+
+FdManager::~FdManager()
+{
+}
+
+auto FdManager::register_handler(mir::Fd fd, std::function<void(int)> const& handler)
+-> std::unique_ptr<FdHandle>
+{
+    auto handle = std::make_unique<FdHandle>(shared_from_this());
+
+    if (auto const main_loop = weak_main_loop.lock().get())
+    {
+        main_loop->register_fd_handler({fd}, &handle, handler);
+    }
+    else
+    {
+        auto fd_info = FdInfo{fd, handle.get(), handler};
+        backlog.push_back(fd_info);
+    }
+
+    return handle;
+}
+
+void FdManager::unregister_handler(void const* owner)
+{
+    if (auto const main_loop = weak_main_loop.lock())
+    {
+        main_loop->unregister_fd_handler(owner);
+    }
+    else
+    {
+        // Remove all entries with same owner from backlog
+        backlog.erase(remove_if(begin(backlog), end(backlog), [&](auto& info){ return info.owner == owner; }), end(backlog));
+    }
+}
+
+void FdManager::process_backlog()
```

Comment:
Why not roll `process_backlog()` into `set_weak_main_loop()`, considering they can be thought of as a single operation? This way the function could take a `shared_ptr`, not have to check if it's null, use it within the function body and then store it as a weak pointer.